src/main/java/io/vertx/spi/cluster/zookeeper/impl/SubsMapHelper.java

```diff
@@ -61,54 +62,84 @@ public void close() {
   }
 
   public void put(String address, RegistrationInfo registrationInfo, Promise<Void> promise) {
-    try {
-      Buffer buffer = Buffer.buffer();
-      registrationInfo.writeToBuffer(buffer);
-      curator.create().orSetData().creatingParentsIfNeeded().withMode(CreateMode.EPHEMERAL).inBackground((c, e) -> {
-        if (e.getType() == CuratorEventType.CREATE || e.getType() == CuratorEventType.SET_DATA) {
-          vertx.runOnContext(Avoid -> {
-            ownSubs.compute(address, (add, curr) -> {
-              Set<RegistrationInfo> res = curr != null ? curr : new CopyOnWriteArraySet<>();
-              res.add(registrationInfo);
-              return res;
+    if (registrationInfo.localOnly()) {
+      localSubs.compute(address, (add, curr) -> addToSet(registrationInfo, curr));
+      fireRegistrationUpdateEvent(address, registrationInfo);
+      promise.complete();
+    } else {
+      try {
+        Buffer buffer = Buffer.buffer();
+        registrationInfo.writeToBuffer(buffer);
+        curator.create().orSetData().creatingParentsIfNeeded().withMode(CreateMode.EPHEMERAL).inBackground((c, e) -> {
+          if (e.getType() == CuratorEventType.CREATE || e.getType() == CuratorEventType.SET_DATA) {
+            vertx.runOnContext(Avoid -> {
+              ownSubs.compute(address, (add, curr) -> addToSet(registrationInfo, curr));
+              promise.complete();
             });
-            promise.complete();
-          });
-        }
-      }).withUnhandledErrorListener(log::error).forPath(fullPath.apply(address, registrationInfo), buffer.getBytes());
-    } catch (Exception e) {
-      log.error(String.format("create subs address %s failed.", address), e);
+          }
+        }).withUnhandledErrorListener(log::error).forPath(fullPath.apply(address, registrationInfo), buffer.getBytes());
+      } catch (Exception e) {
+        log.error(String.format("create subs address %s failed.", address), e);
+      }
     }
   }
 
+  private Set<RegistrationInfo> addToSet(RegistrationInfo registrationInfo, Set<RegistrationInfo> curr) {
+    Set<RegistrationInfo> res = curr != null ? curr : Collections.synchronizedSet(new LinkedHashSet<>());
+    res.add(registrationInfo);
+    return res;
+  }
+
   public List<RegistrationInfo> get(String address) {
-    return Optional.ofNullable(treeCache.getCurrentChildren(keyPath.apply(address))).map(data -> data.values().stream().map(childData -> {
-      RegistrationInfo registrationInfo = new RegistrationInfo();
-      Buffer buffer = Buffer.buffer(childData.getData());
-      registrationInfo.readFromBuffer(0, buffer);
-      return registrationInfo;
-    }).collect(Collectors.toList())).orElse(new ArrayList<>());
+    List<RegistrationInfo> registrationInfos = Optional.ofNullable(localSubs.get(address))
```

Comment:
If you stick with Optional/Stream API here, perhaps you could use Optional<Stream> in both blocks, use concat at the end and let the JDK decide how to create the array list?

That would avoid creating a first list and then re-sizing the array